art/estimators/certification/deep_z/pytorch.py

```diff
@@ -178,9 +142,125 @@ def forward(self, cent: np.ndarray, eps: np.ndarray) -> Tuple["torch.Tensor", "t
             if self.reshape_op_num == op_num:
                 x = x.reshape((x.shape[0], -1))
             x = op(x)
-
         return x[0, :], x[1:, :]
 
+    def concrete_forward(self, in_x: Union[np.ndarray, "torch.Tensor"]) -> "torch.Tensor":
+        """
+        Do the forward pass using the concrete operations
+
+        :param in_x: regular (concrete) data.
+        """
+
+        if isinstance(in_x, np.ndarray):
+            x = torch.from_numpy(in_x.astype("float32")).to(self.device)
+        else:
+            x = in_x
+
+        for op_num, op in enumerate(self.ops):
+            # as reshapes are not modules we infer when the reshape from convolutional to dense occurs
+            if self.reshape_op_num == op_num:
+                x = x.reshape((x.shape[0], -1))
+            x = op.concrete_forward(x)
+        return x
+
+    def set_forward_mode(self, mode: str) -> None:
+        """
+        Helper function to set the forward mode of the model
+
+        :param mode: either concrete or abstract signifying how to run the forward pass
+        """
+        assert mode in {"concrete", "abstract"}
+        self.forward_mode = mode
+
+
+class PytorchDeepZ(PyTorchClassifier, ZonoBounds):
+    """
+    Implementation of DeepZ to certify neural network robustness. We use the zonotope representation of a datapoint as
+    it travels through the network to then verify if it can have its class changed given a certain perturbation.
+
+    | Paper link: https://papers.nips.cc/paper/2018/file/f2f446980d8e971ef3da97af089481c3-Paper.pdf
+    """
+
+    estimator_params = PyTorchClassifier.estimator_params
+
+    def __init__(
+        self,
+        model: "torch.nn.Module",
+        loss: "torch.nn.modules.loss._Loss",
+        input_shape: Tuple[int, ...],
+        nb_classes: int,
+        optimizer: Optional["torch.optim.Optimizer"] = None,  # type: ignore
```

Comment:
Why is this type ignore required? It seems to receive an expected type.

Reply:
Removed: was outdated.